ls1intum/Artemis
# `General`: Fix exercise category selection

### Checklist

#### General

#### Client

- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified
  colors only in the theming variable files and checked that the changes look consistent in both the light and the dark
  theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following
  the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).

### Motivation and Context

Fixes #7449 and further issues that I encountered when working on the issue.

### Description

* Fixed loading of categories when importing exercises (was broken for text, modeling, and programming exercises)
* Fixed the color selection
    * Preventing that click event is emitted twice and interpreted as close event _(caused the issue that the colorpicker could not be opened)_
    * Adjust the logic for 'click-outside' to be more robust
    * Fixed "jumping" of color picker button
* Fixed transparent categories background on selection with arrow keys _(light and dark theme)_
* Fixed reactivity, when removing a category it was not displayed in the selection anymore (Text, Modeling, Quiz, FileUpload)
* Fixed placement of color picker for programming exercises
* Refactorings
    * Added utils function to reduce code duplication for scrollToTopOfPage
    * Added utils function to reduce code duplication for loading the course exercise categories
    * Refactored code for loading supported programming languages _(programming exercise creation)_

### Steps for Testing
Prerequisites:
- 1 Instructor

1. Create / Edit / Import an exercise
2. Adjust the categories of the exercise
3. See that the categories that are already present in the course can be used
4. See that the color picker opens and can be used when clicking on a category
5. See that when removing a category it can be selected again _(just disappeared from selection before)_
6. Verify that Programming Languages can Still be selected when creating a new programming exercise _(e.g. on ts1 java, python, c, haskell, kotlin, vhdl, assembler, swift, ocaml and empty should be available)_

### Testserver States

> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked

![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)

### Review Progress

#### Code Review


#### Manual Tests


### Test Coverage

#### Client

| Class/File | Line Coverage |  Confirmation (assert/expect) |
|------------|--------------:|------------------------------:|
| [exam-result-summary.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5671/latest/artifact/shared/Coverage-Report-Client-Tests/app/exam/participate/summary/exam-result-summary.component.ts.html) | 91.35% |                             ✅ |
| [file-upload-exercise-update.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5671/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts.html) | 85.56% |                             ✅ |
| [example-modeling-submission.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5671/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/modeling/manage/example-modeling/example-modeling-submission.component.ts.html) | 92.24% |                             ✅ |
| [modeling-exercise-update.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5671/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/modeling/manage/modeling-exercise-update.component.ts.html) | 88.18% |                             ✅ |
| [programming-exercise-update.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5671/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/update/programming-exercise-update.component.ts.html) | 80.32% |                             ✅ |
| [programming-exercise-information.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5671/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/update/update-components/programming-exercise-information.component.ts.html) | 100% |                             ❌ |
| [quiz-exercise-detail.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5671/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/quiz/manage/quiz-exercise-detail.component.ts.html) | 92.07% |                             ✅ |
| [course-utils.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5671/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/shared/course-exercises/course-utils.ts.html) | 84.61% |                             ✅ |
| [text-exercise-update.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5671/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/text/manage/text-exercise/text-exercise-update.component.ts.html) | 88.34% |                             ✅ |
| [category-selector.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5671/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/category-selector/category-selector.component.ts.html) | 97.18% |                             ✅ |
| [color-selector.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5671/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/color-selector/color-selector.component.ts.html) | 100% |                             ✅ |
| [utils.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5671/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/util/utils.ts.html) | 96.29% |                             ✅ |


### Screenshots

#### fixed version
https://github.com/ls1intum/Artemis/assets/63976129/944d8f6b-d7b2-4cff-b4f3-47a70601c219

#### broken version
* Categories not loaded on import

https://github.com/ls1intum/Artemis/assets/63976129/b3382973-ef56-4a6c-b142-91fb161c7057
* Colorpicker for programming exercises opening somewhere
  ![category-selection-programming-exercise-broken](https://github.com/ls1intum/Artemis/assets/63976129/0d1758be-8ccb-4c5c-94dc-548f638502c6)

* Remove category misaligned and jumping

https://github.com/ls1intum/Artemis/assets/63976129/225a413f-d95b-4b8d-a1dd-b014ad5bddeb

* Selection becomes transparent when using up and down arrows

![image](https://github.com/ls1intum/Artemis/assets/63976129/f4179086-aba8-446b-b461-74f72d9cc3e1)

* Removed category not available in selection

https://github.com/ls1intum/Artemis/assets/63976129/8d6419c0-e988-43c6-85b2-f933da2fe27a
